crates/wasi-common/src/sys/unix/bsd/hostcalls_impl.rs

```diff
@@ -57,86 +57,102 @@ pub(crate) fn path_symlink(old_path: &str, resolved: PathGet) -> Result<()> {
     log::debug!("path_symlink old_path = {:?}", old_path);
     log::debug!("path_symlink resolved = {:?}", resolved);
 
-    unsafe { symlinkat(old_path, resolved.dirfd().as_raw_fd(), resolved.path()) }.or_else(|err| {
-        if let YanixError::Errno(errno) = err {
-            match errno {
-                Errno::ENOTDIR => {
-                    // On BSD, symlinkat returns ENOTDIR when it should in fact
-                    // return a EEXIST. It seems that it gets confused with by
-                    // the trailing slash in the target path. Thus, we strip
-                    // the trailing slash and check if the path exists, and
-                    // adjust the error code appropriately.
-                    let new_path = resolved.path().trim_end_matches('/');
-                    if let Ok(_) = unsafe {
-                        fstatat(
-                            resolved.dirfd().as_raw_fd(),
-                            new_path,
-                            AtFlag::SYMLINK_NOFOLLOW,
-                        )
-                    } {
-                        Err(Error::EEXIST)
-                    } else {
-                        Err(Error::ENOTDIR)
+    match resolved.dirfd() {
+        Descriptor::OsHandle(file) => {
+            unsafe { symlinkat(old_path, file.as_raw_fd(), resolved.path()) }.or_else(|err| {
+                if let YanixError::Errno(errno) = err {
+                    match errno {
+                        Errno::ENOTDIR => {
+                            // On BSD, symlinkat returns ENOTDIR when it should in fact
+                            // return a EEXIST. It seems that it gets confused with by
+                            // the trailing slash in the target path. Thus, we strip
+                            // the trailing slash and check if the path exists, and
+                            // adjust the error code appropriately.
+                            let new_path = resolved.path().trim_end_matches('/');
+                            if let Ok(_) = unsafe {
+                                fstatat(file.as_raw_fd(), new_path, AtFlag::SYMLINK_NOFOLLOW)
+                            } {
+                                Err(Error::EEXIST)
+                            } else {
+                                Err(Error::ENOTDIR)
+                            }
+                        }
+                        x => Err(x.into()),
                     }
+                } else {
+                    Err(err.into())
                 }
-                x => Err(x.into()),
-            }
-        } else {
-            Err(err.into())
+            })
+        }
+        Descriptor::VirtualFile(_) => {
+            unimplemented!("virtual path_symlink");
```

Comment:
I can see a lot of code duplication between Linux, BSD and Windows hosts here. Is it intentional that we need to handle `Descriptor::VirtualFile` in sys specific `sys::...::hostcalls_impl`? I would imagine the codebase would be easier to maintain if all `Descriptor::VirtualFile` specific were either handle at the sys-common level (in this case, `crate::hostcalls_impl`) *or* perhaps introducing `VirtualFS` as a separate "host" would make sense where we would mimick the structure of sys-specific hostcall functions. I reckon this would make *any* changes to sys-specific functions unnecessary thus minimising the possibility of introducing a bug into any one of them, not to mention the fact, it would simplify our test structure (i.e., as I already mentioned, we wouldn't have to repeat `VirtualFS` tests on every single host).

Reply:
> Is it intentional that we need to handle `Descriptor::VirtualFile` in sys specific `sys::...::hostcalls_impl`?
It's "intentional" in that we'll want to clearly reject the case of `Descriptor::VirtualFile` in `sys::...::hostcalls_impl`, but not intentional in that I don't love delegating to `virtfs` this far into `wasi-common`.

Generally I followed the type system in figuring out where to `match`  have had `OsFile`, I think consistently moving `virtfs` operations to be before any `sys` would make a lot of sense.

However, I _don't_ think we'd want `VirtualFs` as a distinct host - I imagine it wouldn't be too uncommon to be able to serve some files from a local disk and others from something more complex. I'm also not really sure how a `virtfs` host would be selected, are you imagining a `wasi-common` cargo feature?

Reply:
Update: I've moved checks for `Descriptor::VirtualFile` to be `sys`-agnostic, except in `as_os_handle` where it's a clear error and in Windows IO event handling, where the OS delivering an event for a virtual file's fd would be a really surprising error. Since `as_os_handle` will panic when converting a `VirtualFile`, if something goes wrong and they get to `sys`-level implementations, it'll blow up rather than doing something subtly wrong.

This isn't a whole answer to "should virtfs and non-virtfs even be able to mingle?", but assuming the answer is yes it's better than before.

Reply:
Awesome, thanks! IMHO, this looks a lot better now :-)